File: bin/reports_to_xls_v2.rb
#!/bin/env ruby

require 'optparse'
require 'ostruct'
require 'json'
require 'rubyXL'
require 'rubyXL/convenience_methods/cell'
require 'rubyXL/convenience_methods/color'
require 'rubyXL/convenience_methods/font'
require 'rubyXL/convenience_methods/workbook'
require 'rubyXL/convenience_methods/worksheet'

### Define modules and classes here

def parse_json(filename)

    return JSON.parse(IO.readlines(filename).join)

end

### Get the script arguments and open relevant files
options = OpenStruct.new()
opts = OptionParser.new()
opts.banner = "Reads reports and makes an Excel table with one sheet per analysis rule"
opts.separator ""
opts.on("-o","--outfile", "=OUTFILE","Output file") {|argument| options.outfile = argument }
opts.on("-h","--help","Display the usage information") {
    puts opts
    exit
}

opts.parse! 

color = {
	"even" => "FFFFFF",
	"uneven" => "d4e6f1"
}

negative_result = "-"

files = Dir["*.json"]

toolchains = []

bucket = {}

files.each do |file|

    json        = parse_json(file)
    sample      = json["sample"]
    matches     = json["matches"]

    matches.each do |match|

        toolchain   = match["toolchain"]
        rule        = match["rule"]

        toolchains << toolchain unless toolchains.include?(toolchain)

        bucket[rule] = {} unless bucket.has_key?(rule)

        bucket[rule].has_key?(sample) ? bucket[rule][sample] << match : bucket[rule][sample] = [ match ]

    end

end

toolchains.sort!

workbook = RubyXL::Workbook.new
page = 0

# a bucket is a rule with all the matching reports, i.e. one page
bucket.each do |rule,samples|

    sheet = workbook.worksheets[page]
    sheet.sheet_name = rule

    row = 0
    col = 0

    [ "" ].push(toolchains.map { |tc| [ tc, "", ""] }).flatten.each do |tc|
        sheet.add_cell(row,col,tc)
        sheet.sheet_data[row][col].change_font_bold(true)
        sheet.change_column_width(col, 15)
        col += 1
    end
    row += 1
    col = 0

    [ "Probe" ].push(toolchains.map{|tc| [ "% GMO","Reads WT","Reads GMO"]}).flatten.each do |tc|
        sheet.add_cell(row,col,tc)
        sheet.sheet_data[row][col].change_font_bold(true)
        sheet.change_column_width(col, 15)
        col += 1
    end

    samples.each do |sample,reports|

        row += 1
        col = 0

        sheet.add_cell(row,col,sample)
        col += 1

        failed = false

        toolchains.each do |tool|

            perc_gmo = ""
            ref_cov = ""
            alt_cov = ""

            report = reports.find{|r| r["toolchain"] == tool }

            if report
                perc_gmo = report["perc_gmo"]
                ref_cov = report["ref_cov"]
                alt_cov = report["alt_cov"]
                if ref_cov == "NA" && report.has_key?("bam_cov")
                    ref_cov = report["bam_cov"]
                    alt_cov = "-"
                end
            end

            # Simplistic rule to catch failed samples. 
            if ref_cov.to_i < 100
                failed = true
            end

            [ perc_gmo, ref_cov, alt_cov].each do |e|
                sheet.add_cell(row,col,e)
                col += 1
            end


        end

        if failed
            sheet.change_row_fill(row,"FF3300")
        else
            row.even? ? bg = color["even"] : bg = color["uneven"]
            sheet.change_row_fill(row, bg)
        end
        sheet.change_row_horizontal_alignment(row, 'right')

    end
    
    col = 0
    toolchains.each do |tc|
        sheet.change_column_border(col, :right, 'medium')
        col += 3
    end
    
    sheet.change_row_border(0, :bottom, 'medium')  

    # increment page counter for the next rule, if any
    page += 1

end

workbook.write(options.outfile)

File: bin/reports_to_xls_v2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


import argparse
import glob
import json
from openpyxl import Workbook
from openpyxl.worksheet.dimensions import ColumnDimension, DimensionHolder
from openpyxl.utils import get_column_letter
from openpyxl.styles import Font, PatternFill

parser = argparse.ArgumentParser(description="Script options")
parser.add_argument("--input", help="An input option")
parser.add_argument("--output")
args = parser.parse_args()

def flatten_extend(matrix):
    flat_list = []
    for row in matrix:
        if type(row) is list:
            flat_list.extend(row)
        else:
            flat_list.append(row)
    return flat_list


def main(output):

    wb = Workbook()
    ft = Font(name="Sans", bold=True)
    cb_even = PatternFill(fill_type="solid", fgColor="d9e1f2")
    cb_uneven = PatternFill(fill_type="solid", fgColor="cdd1d9")
    cb_failed = PatternFill(fill_type="solid", fgColor="FF3300")

    toolchains = []
    bucket = {}

    reports = sorted(glob.glob("*.json"))

    # All the JSON reports
    for report in reports:

        # Parse the JSON file
        with open(report) as f:
            data = json.load(f)

        sample = data["sample"]
        matches = data["matches"]

        # group by rule, then sample
        for match in matches:
            toolchain = match["toolchain"]
            rule = match["rule"]

            if not toolchain in toolchains:
                toolchains.append(toolchain)

            if rule not in bucket:
                bucket[rule] = {}

            if sample in bucket[rule]:
                bucket[rule][sample].append(match)
            else:
                bucket[rule][sample] = [ match ]

    toolchains.sort()

    sheet_index = 0

    # write one sheet per rule, listing all the samples
    for rule, samples in bucket.items():
         
        ws = wb.create_sheet(title=rule, index=sheet_index)
        sheet_index += 1

        header = [ "" ] 
        for tool in toolchains:
            header.append( [tool, "", ""])
        ws.append(flatten_extend(header))

        header = [ "Probe" ]
        for tool in toolchains:
            header.append([ "% GMO","Reads WT","Reads GMO"])

        ws.append(flatten_extend(header))

        row = 0

        for sample, reports in samples.items():

            row += 1

            this_row = [ sample ]

            failed = False

            for tool in toolchains:

                perc_gmo = ""
                ref_cov = ""
                alt_cov = ""

                report = [d for d in reports if d['toolchain'] == tool ][0]

                if report:
                    perc_gmo = float(report["perc_gmo"])
                    ref_cov = int(report["ref_cov"])
                    alt_cov = int(report["alt_cov"])
                    if ref_cov == "NA" and "bam_cov" in report:
                        ref_cov = report["bam_cov"]
                        alt_cov = "-"
                
                # Simplistic rule to catch failed samples. 
                if ref_cov < 100:
                    failed = True

                for e in [ perc_gmo, ref_cov, alt_cov ]:
                    this_row.append(e)

            ws.append(this_row)

            if failed:
                bg_color = cb_failed
            else:
                bg_color = cb_even if (row & 1) else cb_uneven
                
            for col in ["A", "B", "C", "D", "E", "F", "G", "H"]:
                ws[col+str(ws._current_row)].fill = bg_color       

         # Auto-width for columns
        dim_holder = DimensionHolder(worksheet=ws)
        for column in range(ws.min_column, ws.max_column + 1):
            dim_holder[get_column_letter(column)] = ColumnDimension(ws, min=column, max=column, width=20)
        ws.column_dimensions = dim_holder
        ws.freeze_panes = ws["A3"]

    # Write excel file
    wb.save(output)
    

if __name__ == '__main__':
    main(args.output)
